src/compiler/scala/tools/nsc/PhaseAssembly.scala

```diff
@@ -224,38 +234,92 @@ object DependencyGraph {
-      for (p <- phases) {
-        require(p.phaseName.nonEmpty, "Phase name must be non-empty.")
-        def checkConstraint(name: String, constraint: String): Boolean =
-          graph.components.contains(name).tap(ok => if (!ok) graph.warning(s"No phase `$name` for ${p.phaseName}.$constraint"))
-        for (after <- p.runsRightAfter if after.nonEmpty && checkConstraint(after, "runsRightAfter"))
-          graph.addEdge(after, p.phaseName, FollowsNow)
-        for (after <- p.runsAfter if after.nonEmpty && !p.runsRightAfter.contains(after) && checkConstraint(after, "runsAfter"))
-          graph.addEdge(after, p.phaseName, Follows)
-        for (before <- p.runsBefore if before.nonEmpty && checkConstraint(before, "runsBefore"))
-          graph.addEdge(p.phaseName, before, Follows)
-        if (p != start && p != end)
-          if (p.runsRightAfter.forall(_.isEmpty) && p.runsAfter.forall(_.isEmpty))
-            graph.addEdge(start.phaseName, p.phaseName, Follows)
-        if (p != end || p == end && p == start)
-          if (p.runsBefore.forall(_.isEmpty))
-            graph.addEdge(p.phaseName, end.phaseName, Follows)
+    val constraints = ListBuffer.empty[(String, String, Weight)]
+    def addConstraint(from: String, to: String, weight: Weight): Unit = constraints.addOne((from, to, weight))
+    val warnings = ListBuffer.empty[String]
+    var i = 0
+    while (i < phases.length) {
+      val p = phases(i)
+      require(p.phaseName.nonEmpty, "Phase name must be non-empty.")
+      var constrained = false
+      var badConstraint = false
+      def checkConstraint(name: String, constraint: String, warn: Boolean): Boolean = isPhaseName(name).tap { ok =>
+        if (!ok && warn) {
+          val help = if (compilerPhases.contains(name)) "" else
+            compilerPhases.filter(util.EditDistance.levenshtein(name, _) < 3) match {
+              case Nil => ""
+              case close => s" - did you mean ${util.StringUtil.oxford(close, "or")}?"
+            }
+          warnings.addOne(s"No phase `$name` for ${p.phaseName}.$constraint$help")
+          badConstraint = true
+        }
       }
+      // add edges for each constraint
+      for (after <- p.runsRightAfter if after.nonEmpty && checkConstraint(after, "runsRightAfter", warn=warnAll)) {
+        addConstraint(after, p.phaseName, FollowsNow)
+        constrained = true
+      }
+      for (after <- p.runsAfter if after.nonEmpty && !p.runsRightAfter.contains(after) && checkConstraint(after, "runsAfter", warn=warnAll)) {
+        addConstraint(after, p.phaseName, Follows)
+        constrained = true
+      }
+      for (before <- p.runsBefore if before.nonEmpty && checkConstraint(before, "runsBefore", warn=warnAll)) {
+        addConstraint(p.phaseName, before, Follows)
+        constrained = true
+      }
+      // if it doesn't follow a phase in the run, replace this component with a no-op that follows start
```

Comment:
If there are two phases where `p2 runsAfter p1` and `p1 runsAfter typo`, then only `p1` is replaced by a noop.

Here's another approach, I'll clean it up and adjust / add tests: https://github.com/scala/scala/compare/2.13.x...lrytz:scala:pr10856?expand=1

Reply:
I changed the noop to a list of names to filter out.